### PR TITLE
lazy val check

### PR DESCRIPTION
**Pull request by nitay**
@MasterDDT caught a problem when we cleaned up our DI where the chunk poller wasn't be initialized right. This is because it has a val and therefore the order in which it was being included became important. I believe this is the same problem we have seen with AiqLogger. The fix is to always use `lazy val` inside traits, then the order you mix them in doesn't matter.

This scalastyle rule checks for that.

Look at `trait_test.sc` for an example of what can happen.
Check out the spec, let me know if you guys think of some other tricky cases I should add. Also I'm going to run this on AIQ codebase - that will help me fix any false positives, but it won't tell me about false negatives so if you think of some tricky thing that might fool and pass this lmk.

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/ActionIQ/scalastyle-rules/pull/2" target="_blank">ReviewNinja

**Comment by MasterDDT** [comment]
Lazy vals are getting cheaper: http://docs.scala-lang.org/sips/pending/improved-lazy-val-initialization.html. Also I haven't done a CPU/timing profile of our page browse, but I'd be willing to bet we could hit 100+ lazy vals and that still wouldn't be a big chunk of where we spend our time. 

So I'd say lazy all the vals, its worth saving the time investigating the errors. Its extremely hard even when its your own code, because you wont know it. I had to make a change to the `ChunkPoller` because I slicked the `SegmentDefinitionDao`...that not a nice stable codebase.

But maybe decorate it somehow so we know which vals were made lazy by default vs the developer intended it? And maybe it is worth a CPU/timing profile?

**Comment by MasterDDT** [comment]
!star

File: src/test/scala/co/actioniq/scalastyle/TraitValShouldBeLazySpec.scala [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+class TraitValShouldBeLazySpec extends Specification with CheckerTestHelper {
+  val key = TraitValShouldBeLazy.errorKey
+  val classUnderTest = classOf[TraitValShouldBeLazy]
+
+  "TraitValShouldBeLazy" should {
+    "error on missing lazy" in {
+      val source = """
+                    |trait Foobar {
+                    |  val x0 = true
```
Comment (MasterDDT) [inline src/test/scala/co/actioniq/scalastyle/TraitValShouldBeLazySpec.scala]: Is it worth trying `val x0 = someOtherVal + someDef`? I guess its just the left side that matters?

File: src/test/scala/co/actioniq/scalastyle/TraitValShouldBeLazySpec.scala [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                     |    val x0 = true
+                     |  }
+                     |}
+                   """.stripMargin
+      assertErrors(List(columnError(4, 4)), source)
+    }
+
+    "error in trait inside trait" in {
+      val source = """
+                     |trait Box {
```
Comment (MasterDDT) [inline src/test/scala/co/actioniq/scalastyle/TraitValShouldBeLazySpec.scala]: How about `object Box { object InnerBox { val x0 = true`?

**Comment by nitay** [comment]
not sure how to decorate it, what do you have in mind?
I'm also thinking I'll add to this (or make a separate rule) check that a val inside a trait is final

File: src/test/scala/co/actioniq/scalastyle/TraitValShouldBeLazySpec.scala (hunk 1 above)
Comment (nitay) [inline src/test/scala/co/actioniq/scalastyle/TraitValShouldBeLazySpec.scala]: will add for completeness sake

File: src/test/scala/co/actioniq/scalastyle/TraitValShouldBeLazySpec.scala (hunk 2 above)
Comment (nitay) [inline src/test/scala/co/actioniq/scalastyle/TraitValShouldBeLazySpec.scala]: should be covered by ""no error in object" but I can add one for this if you like

**Pull request merged**